### ISSUE TITLE
Design and implementation of the Caching layer

### ISSUE THREAD

**Issue by renatomassaro** [HackerExperience/Helix#230]
# Design and implementation of the Caching layer

Implementation of a Caching layer has been postponed as much as we could, but it's time to actually do it. It will simplify APIs a lot, and also save load on our production servers (currently, we are only concerned with the former).

We store data on a highly normalized fashion (3NF for most models). This is good for writes, but not so much for reads.

Joins aren't the only problem... In addition to the normalization, we also store the data in different *databases*, one for each domain. This is, er, by [design](https://dev.hackerexperience.com/w/hacker_experience/architecture/).

In short, we play around with the CAP theorem. Within domains, we ensure a high level of consistency. Between domains, we have only eventual consistency. This model works fairly well for our use case.

Of course, designing the database in such a way brings disadvantages that we need to work around. One of them is a high level of "query chaining" between intermediary systems.

For example, suppose we want to find out the owner of an IP address. Currently, we must:

 1) get the network_connection id; 2) get the NIC component; 3) get the motherboard slot the NIC is linked to; 4) get the motherboard_id; 5) get the server_id the motherboard is attached to; 6) get the entity which owns that server.

Queries 1 to 4 happen within the Hardware domain. Query 5 happens on the Server domain. And query 6 happens on the Entity domain.

The caching abstraction I propose here is intended to solve this unnecessary chaining of queries when reading data, while maintaining the benefits of a highly normalized data model.

## The Cache service

`Cache` is a service within Helix. It has its own database, models, API etc, like any other service.

It is the only entity responsible for reading and writing to the Cache database. It is also responsible for cleaning up expired entries, as well as constantly scanning for inconsistencies.

When looking up an entry, the cache may be found (hit) or not (miss). If it's a hit, the data is returned. If it's a miss, the cache is populated and then another lookup is performed.

Cache invalidation is informed asynchronously by the service that owns the corresponding data, using our event-based architecture.

This means Cache is eventually consistent. It is possible the Cache service will reply with incorrect data up to 50ms *after* the actual change[1]. These trade-offs are inherent of distributed systems and it's acceptable within our application.

Do note that there are ways to minimize, and essentially eliminate, serving of potentially stale data. However, it's a lot of trouble for little gain. For now, we'll accept this. If we find out that it is actually impairing with the gameplay, we'll make changes to increase consistency of Cache.

### Expired entries

We should probably set an expiration time to all cached entries in order to 1) eliminate unused data; 2) decrease storage usage; 3) remove potentially inconsistent data; 4) keep cache lookups fast.

We could set a default expiration time, like 24 hours after insertion, but I believe we should allow some tables to have their own expiration interval. Some data changes faster than others.

The expiration time is renewed after updates.

### Scanning for inconsistencies

When the system load is not too high, we could perform automatic queries that will scan the data source and ensure the cache is up-to-date.

I believe once we have production data we'll be able to spot "hot tables" that are more prone to error than others. In a way, this could be seen as regression testing.

Of course, all in good time. Implementation of inconsistency scanning can and should be postponed. For future reference, something like [pg_cron](https://github.com/citusdata/pg_cron) comes to mind.

### Datastore

I've always meant to use Aerospike as our Cache datastore, specially for the "sub-ms response time" advertised (to be honest it *is* insanely fast). However I want to speed things up, and deploying and maintaining yet another server(s) is a temporary no-no.

As a result, I plan to use our good old Postgresql server to store Cache data. It's fast enough. If we do hit any performance bottleneck, migrating to Aerospike shouldn't be a problem.

### Data model

Naturally, we don't need to assert any kind of referential integrity at the data level. This may marginally increase write time.

Cache data should be denormalized as much as possible. The best design for efficient look-ups is to keep related data in the same table, so only one query at one table can serve multiple requests.

### Proactive caching

We can proactively populate the cache *before* the user attempts to read from it. Some usage patterns / behaviors are predictable and we should take advantage of it. For instance, when logging in, it's quite likely that the user will fetch a lot of data from herself, so it's probably a good idea to cache all her servers, storages, files, hardware etc.

### Future improvements

We could add one other Cache layer in the Erlang cluster, using something like [dets](http://erlang.org/doc/man/dets.html). Definitely not worth the trouble now, but it's an option. As they say, [there is no problem in computer science which cannot be solved by one more level of indirection](http://wiki.c2.com/?OneMoreLevelOfIndirection)

---

[1] - 50ms is a "safe" approach within the 95th percentile of all requests. Under extreme scenarios, it is possible data will be stale for seconds up to a couple minutes.

**Comment by renatomassaro** [comment]
There's one important gotcha we must be aware of:

Since cache purges are asynchronous by design, it's very important that no subsequent cache queries are made within the same request/process

New requests coming in afterwards are ok: when they arrive, the purge *should* have been executed already. Otherwise, this is a source of subtle and hard-to-debug bugs.

I have one idea to increase the system's resilience in this regard. It's a simple approach to what was discussed previously ("*there are ways to minimize, and essentially eliminate, serving of potentially stale data.*"), meaning it's neither too complex nor fail-safe. The idea is:

- Store in cluster memory a list/hash of cache entries queued up for deletion.
- When looking up, check whether that entry/hash isn't queued for deletion.
- Once purged, remove entry/hash from memory.

In fact, if we make a more robust queue, we can delay and sync cache purges every few minutes or so, reducing load of the system. Initially, however, I think we should stick to the simplest solution.

It's simple, but still eventually consistent and prone to failure, for instance, on (rare) node restarts. Yet, it brings a good benefit that we can make subsequent queries on the same request without worrying about stale data. 

## Implementation of the Purge Queue

It would be trivial if we were talking about a single node, just use ETS and call it a day. Even though we will use very few nodes, we still need to support n>1, which makes the queue implementation extremely more difficult.

I think the sanest way to implement a simple queue would be to rely on Mnesia to do the hard work.

Now, my biggest gripe with Mnesia (or, the trade-offs it makes) is about split-brains from potential network partitions. I want to free myself from the trouble of handling these partitions. In this case, I think we should:

- guarantee that requests by the same person are always served by the same client (stick requests).
- in case of split brain without auto merge, *drop all data and start afresh*.
  - for the purge queue, start a new empty queue while attempting to delete the old values.

The above two conditions can only happen if we adopt similar philosophy for the entire cluster. There are other Helix systems that needs distributed coordination (TOP, Phoenix Channels comes to mind), so we ( @mememori ) should discuss the best way to solve these problems having in mind that:

- We'll have small clusters, probably never going anywhere north of n=5.
- Most or all distributed requirements are for cached data that can be recalculated/derived from the DB.

In short, I think we should make each node "independent", based on requests stickness, and eventually replicate in-memory data. So if a node goes down, we recalculate requests hashes, redirecting them to a different node. We've had this conversation a while ago. Now we just have to write it down.

**Comment by renatomassaro** [comment]
It's there
